### PR TITLE
[ESIMD] Add support for vectorizing scalar functions - FE part

### PR DESCRIPTION
**Pull request by DenisBakhvalov**
This patch is one of the few patches to provide a functionality to reuse
existing scalar function implementations in a vector context.

For example, based on scalar version of `float cos(float)`, implement
simd<float,N> cos(simd<float,N>) - "widened" version of the original
scalar function.

To achieve that, we mark such functions with `sycl_esimd_widen` attribute
which will be later used by vector compiler backend to widen it.

For now, this functionality will be used only in internal implementation of
specific public ESIMD APIs. It is not supposed to be used directly in the
user code. It is a temporary solution to enable ESIMD math functions by
reusing existing scalar implementations. Later it is supposed to be
replaced by a more generic mechanism.

### CODE REVIEW

File: clang/lib/CodeGen/CGSYCLRuntime.cpp [hunk 1]
```diff
@@ -64,6 +64,10 @@ bool CGSYCLRuntime::actOnFunctionStart(const FunctionDecl &FD,
   if (FD.hasAttr<SYCLSimdAttr>())
     F.setMetadata("sycl_explicit_simd", llvm::MDNode::get(F.getContext(), {}));
 
+  // Set the function attribute expected by the vector backend compiler.
+  if (const SYCLIntelESimdWidenAttr *A = FD.getAttr<SYCLIntelESimdWidenAttr>())
```
Comment (premanandrao) [inline clang/lib/CodeGen/CGSYCLRuntime.cpp:68]: You can make this auto.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 2]
```diff
@@ -5087,6 +5087,27 @@ static void handleSYCLRegisterNumAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
   D->addAttr(::new (S.Context) SYCLRegisterNumAttr(S.Context, AL, RegNo));
 }
 
+static void handleSYCLIntelESimdWidenAttrAttr(Sema &S, Decl *D,
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaDeclAttr.cpp:5090]: Can you please rewrite this to use 'Add' and 'Merge' methods. You can see https://github.com/intel/llvm/pull/3224 for examples. This aligns with community standards better and will help us with our upstreaming efforts.

Also, I don't think redeclarations are currently handled in this code i.e. cases like - 
```
[[intel::sycl_esimd_widen(8)]] void func();
[[intel::sycl_esimd_widen(16)]] void func() {}
```
You can add a 'Merge' method for this.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 3]
```diff
@@ -5087,6 +5087,27 @@ static void handleSYCLRegisterNumAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+                           /*StrictlyUnsigned=*/true))
+    return;
+
+  if (NumSimdLanes != 8 && NumSimdLanes != 16 && NumSimdLanes != 32) {
+    S.Diag(Attr.getLoc(), diag::err_sycl_esimd_widen_unsupported_value)
+        << Attr << E->getSourceRange();
+    return;
+  }
+
+  if (D->getAttr<SYCLIntelESimdWidenAttr>())
```
Comment (premanandrao) [inline clang/lib/Sema/SemaDeclAttr.cpp:5104]: check D->hasAttr instead?

File: clang/test/SemaSYCL/esimd_widen.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify %s
```
Comment (premanandrao) [inline clang/test/SemaSYCL/esimd_widen.cpp:1]: Can you add a check for the host case too?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 5]
```diff
@@ -5087,6 +5087,27 @@ static void handleSYCLRegisterNumAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
   D->addAttr(::new (S.Context) SYCLRegisterNumAttr(S.Context, AL, RegNo));
 }
 
+static void handleSYCLIntelESimdWidenAttrAttr(Sema &S, Decl *D,
+                                              const ParsedAttr &Attr) {
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:5091]:
```suggestion
                                              const ParsedAttr &PA) {
```
Or any other identifier that isn't a type name.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 3 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:5104]: FWIW, this won't catch the issue on redeclarations (see the comment above from @elizabethandrews).

File: clang/test/SemaSYCL/esimd_widen.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+// This test check the semantics of sycl_esimd_widen attribute
+
+// expected-error@+1{{'sycl_esimd_widen' attribute argument must be 8, 16, or 32}}
+__attribute__((sycl_esimd_widen(17))) void foo1() {}
```
Comment (AaronBallman) [inline clang/test/SemaSYCL/esimd_widen.cpp:6]: Can you also add a test that shows we diagnose when you try to apply the attribute to a non-function or is given a non-integer argument?

File: clang/include/clang/Basic/Attr.td [hunk 7]
```diff
@@ -1239,6 +1239,23 @@ def SYCLSimdAccessorPtr : InheritableAttr {
   let Documentation = [Undocumented];
 }
 
+// The attribute denotes that it is a function written in SIMT fashion, which
```
Comment (kbobrovs) [inline clang/include/clang/Basic/Attr.td:1242]: We should not say that the function is written in SIMT fashion, as SIMT fashion allows workitems to have independent forward progress, in which case we'd require developer to use subgroup barriers for synchronization in some cases. Instead, we assume this function executes each operator in lock-step after vectorization (widening). Please adjust the comment.

File: clang/include/clang/Basic/Attr.td [hunk 8]
```diff
@@ -1239,6 +1239,23 @@ def SYCLSimdAccessorPtr : InheritableAttr {
...
+// is used in ESIMD context and needs to be widened by a vector backend compiler.
+// For now, this attribute will be used only in internal implementation of
+// specific public ESIMD APIs. It is not supposed to be used directly in the
+// user code, hence it is undocumented.
+// The argument of the attribute specifies the number of SIMD lanes, for which
+// the function should be widened.
+def SYCLIntelESimdWiden : InheritableAttr {
+  let Spellings = [GNU<"sycl_esimd_widen">, CXX11<"intel", "sycl_esimd_widen">];
```
Comment (kbobrovs) [inline clang/include/clang/Basic/Attr.td:1250]: "widening" is kind of non-standard term, "vectorization" is. I'd suggest something like "sycl_esimd_auto_vectorize" (other variants - "sycl_esimd_vectorize", "sycl_esimd_vector").

File: clang/include/clang/Basic/Attr.td [hunk 9]
```diff
@@ -1239,6 +1239,23 @@ def SYCLSimdAccessorPtr : InheritableAttr {
...
+  let Args = [UnsignedArgument<"Number">];
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply clang/include/clang/Basic/Attr.td:1251]: Yes, I it should.

[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/include/clang/Basic/Attr.td:1251]: Thanks! In this case, this should be taking an `ExprArgument` instead -- see the implementation of `SYCLIntelNumSimdWorkItems` for a good example of what to model from.

File: clang/include/clang/Basic/Attr.td (hunk 7 above)
Comment (DenisBakhvalov) [reply clang/include/clang/Basic/Attr.td:1242]: I can replace the word "SIMT" with "scalar". Would that be sufficient? I don't think we need to dig too much into the details here.

File: clang/include/clang/Basic/Attr.td (hunk 8 above)
[the remark this answers is not in the document]
Comment (DenisBakhvalov) [reply clang/include/clang/Basic/Attr.td:1250]: I'm ok with just leaving a GNU-style, but is there a downside of supporting both styles?

Comment (DenisBakhvalov) [reply clang/include/clang/Basic/Attr.td:1250]: Ok, I talked to a member of the vectorization team and they suggested that "vectorization" would be indeed better. I will go with `sycl_esimd_vectorize`.

[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/include/clang/Basic/Attr.td:1250]: No major downside outside of the usual annoyance with GNU-style attributes being complex to use (my preference is to use `[[]]` spellings because it's a SYCL-related attribute and SYCL requires C++, so I don't see why the GNU spelling would ever be needed).

[the remark this answers is not in the document]
Comment (DenisBakhvalov) [reply clang/include/clang/Basic/Attr.td:1250]: That makes sense. Will update.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 2 above)
Comment (DenisBakhvalov) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/lib/CodeGen/CGSYCLRuntime.cpp (hunk 1 above)
Comment (DenisBakhvalov) [reply clang/lib/CodeGen/CGSYCLRuntime.cpp]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 3 above)
Comment (DenisBakhvalov) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Reimplemented.

File: clang/test/SemaSYCL/esimd_widen.cpp (hunk 4 above)
Comment (DenisBakhvalov) [reply clang/test/SemaSYCL/esimd_widen.cpp]: Done

File: clang/include/clang/Basic/Attr.td (hunk 9 above)
[the remark this answers is not in the document]
Comment (DenisBakhvalov) [reply clang/include/clang/Basic/Attr.td]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 5 above)
Comment (DenisBakhvalov) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/test/SemaSYCL/esimd_widen.cpp (hunk 6 above)
Comment (DenisBakhvalov) [reply clang/test/SemaSYCL/esimd_widen.cpp]: Done

File: clang/include/clang/Basic/Attr.td (hunk 7 above)
Comment (kbobrovs) [reply clang/include/clang/Basic/Attr.td]: scalar is ok

File: clang/include/clang/Basic/Attr.td (hunk 8 above)
Comment (kbobrovs) [reply clang/include/clang/Basic/Attr.td]: ok

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 10]
```diff
@@ -5087,6 +5087,79 @@ static void handleSYCLRegisterNumAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
+
+    // This attribute requires a strictly positive value.
+    if (ArgVal <= 0) {
```
Comment (kbobrovs) [inline clang/lib/Sema/SemaDeclAttr.cpp:5104]: Nit:
this check is not needed actually, as the check below does needed verification.

**Review by kbobrovs — DISMISSED** [review]
LGTM from my side.

File: clang/test/SemaSYCL/esimd-vectorize-attr-host.cpp [hunk 11]
```diff
@@ -0,0 +1,9 @@
+// RUN: %clang_cc1 -fsycl-is-host -fsyntax-only -verify %s
+
+// This test check the semantics of sycl_esimd_vectorize attribute
+
+// expected-warning@+1{{'sycl_esimd_vectorize' attribute ignored}}
+[[intel::sycl_esimd_vectorize(17)]] void foo1() {}
+
+// expected-warning@+1{{unknown attribute 'sycl_esimd_vectorize' ignored}}
```
Comment (AaronBallman) [inline clang/test/SemaSYCL/esimd-vectorize-attr-host.cpp:8]:
```suggestion

// We explicitly do not support a GNU spelling for this attribute, which is why it is
// treated as an unknown attribute.
// expected-warning@+1{{unknown attribute 'sycl_esimd_vectorize' ignored}}
```

File: clang/test/SemaSYCL/esimd-vectorize-attr-device.cpp [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+// expected-error@+1{{'sycl_esimd_vectorize' attribute requires a positive integral compile time constant expression}}
+[[intel::sycl_esimd_vectorize(-8)]] void foo2() {}
+// expected-error@+1{{integral constant expression must have integral or unscoped enumeration type, not 'float'}}
+[[intel::sycl_esimd_vectorize(3.f)]] void foo3() {}
+
+[[intel::sycl_esimd_vectorize(8)]] void foo4() {}
+[[intel::sycl_esimd_vectorize(16)]] void foo5() {}
+[[intel::sycl_esimd_vectorize(32)]] void foo6() {}
+
+// expected-warning@+1{{unknown attribute 'sycl_esimd_vectorize' ignored}}
```
Comment (AaronBallman) [inline clang/test/SemaSYCL/esimd-vectorize-attr-device.cpp:16]:
```suggestion
// We explicitly do not support a GNU spelling for this attribute, which is why it is
// treated as an unknown attribute.
// expected-warning@+1{{unknown attribute 'sycl_esimd_vectorize' ignored}}
```

**Review by AaronBallman — COMMENTED** [review]
Mostly just nits from me.

**Review by premanandrao — COMMENTED** [review]
LGTM, other than the changes already requested by others.

File: clang/include/clang/Basic/Attr.td (hunk 9 above)
[the remark this answers is not in the document]
Comment (elizabethandrews) [reply clang/include/clang/Basic/Attr.td]: Please add some tests for templates. I don't think it will work at the moment. I think you need to add code in SemaTemplateInstantiateDecl.cpp

File: clang/lib/CodeGen/CGSYCLRuntime.cpp [hunk 13]
```diff
@@ -65,8 +65,12 @@ bool CGSYCLRuntime::actOnFunctionStart(const FunctionDecl &FD,
     F.setMetadata("sycl_explicit_simd", llvm::MDNode::get(F.getContext(), {}));
 
   // Set the function attribute expected by the vector backend compiler.
-  if (const SYCLIntelESimdWidenAttr *A = FD.getAttr<SYCLIntelESimdWidenAttr>())
-    F.addFnAttr("CMGenxSIMT", std::to_string(A->getNumber()));
+  if (const auto *A = FD.getAttr<SYCLIntelESimdVectorizeAttr>())
+    if (const auto *DeclExpr = dyn_cast<ConstantExpr>(A->getValue())) {
```
Comment (elizabethandrews) [inline clang/lib/CodeGen/CGSYCLRuntime.cpp:69]: Since the argument isn't optional I'm not sure if we need this check. I think this can be replaced by an assert @AaronBallman

Comment (AaronBallman) [reply clang/lib/CodeGen/CGSYCLRuntime.cpp:69]: Yup -- the `dyn_cast` should be a `cast` because by CodeGen time we've already resolved any dependent expressions.

File: clang/include/clang/Basic/Attr.td (hunk 9 above)
[the remark this answers is not in the document]
Comment (DenisBakhvalov) [reply clang/include/clang/Basic/Attr.td]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 10 above)
Comment (DenisBakhvalov) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/lib/CodeGen/CGSYCLRuntime.cpp (hunk 13 above)
Comment (DenisBakhvalov) [reply clang/lib/CodeGen/CGSYCLRuntime.cpp]: Done

**Review by AaronBallman — APPROVED** [review]
LGTM!

File: clang/test/CodeGenSYCL/esimd-vectorize-md.cpp [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+// This test checks the generation of CMGenxSIMT function attributes
+
+[[intel::sycl_esimd_vectorize(32)]] __attribute__((sycl_device)) void foo1() {}
+// CHECK: @_Z4foo1v() #[[ATTR1:[0-9]+]]
+
+[[intel::sycl_esimd_vectorize(8)]] [[intel::sycl_esimd_vectorize(16)]] __attribute__((sycl_device)) void foo2() {}
+// CHECK: @_Z4foo2v() #[[ATTR2:[0-9]+]]
+
+[[intel::sycl_esimd_vectorize(8)]] __attribute__((sycl_device)) void foo3();
```
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: This isn't a review comment but just for my own curiosity. I would have expected `8` to be applied here since that is the attribute in first declaration of `foo3()` . Off the top of your head, do you know why `16 `is applied @AaronBallman ?

**Review by elizabethandrews — APPROVED** [review]

File: clang/test/CodeGenSYCL/esimd-vectorize-md.cpp (hunk 14 above)
Comment (AaronBallman) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: That's a good point -- the code in `MergeSYCLIntelESimdVectorizeAttr` should be dropping the second attribute...

Comment (AaronBallman) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: I think I may have confused you in this -- I thought the code *was* using the initial value because when merging, we would return nullptr for the merged attribute. So I thought the test was wrong, not the Sema code.

I think the new changes are wrong (sorry about that) because the warning is saying that the second attribute is ignored.

Comment (DenisBakhvalov) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: The warning says that the attribute is already applied with different arguments. It doesn't say which one is ignored.

So, which one should be selected in this case? `16`? Should I revert my recent change then?

Comment (AaronBallman) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: The location of the diagnostic should make it clear that one is being diagnosed is the second one. See the warning on line 26 of clang/test/SemaSYCL/esimd-vectorize-attr-device.cpp.

I think your most recent change should be reverted and you should debug why CodeGen is seeing the later attribute rather than the earlier one. It may help to look at an -ast-dump of the code just to make sure it looks reasonable.

Comment (DenisBakhvalov) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: So should I use something like `FD.getPreviousDecl()` in `CGSYCLRuntime.cpp` to get this fixed?

From the AST point of view, everything looks as expected: declaration has the attribute with value `8`, definition has the attribute with value `16`.

Comment (DenisBakhvalov) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: @AaronBallman , ping

Comment (AaronBallman) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: Sorry for the delayed response, somehow this review got lost in my queue. :-(

> From the AST point of view, everything looks as expected: declaration has the attribute with value 8, definition has the attribute with value 16.

What I'd expect is that a conflict between values would be an error rather than a warning, to be honest (that's how most attributes work). But if we want it to be a warning instead, I'd expect the AST to look like:

|-FunctionDecl foo3 'void ()'
| `-SYCLIntelESimdVectorizeAttr 8
`-FunctionDecl prev foo3 'void ()'
  `-SYCLIntelESimdVectorizeAttr Inherited 8

e.g., it should inherit the previous attribute and not add a new one.

Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: > What I'd expect is that a conflict between values would be an error rather than a warning, to be honest (that's how most attributes work). 
I think conflicting values currently emit a warning for SYCL attributes. If we want to change this, this should be changed for the existing attributes also right?

Anyway, I haven't verified this but from a quick look I think you need to add a `return nullptr` after diagnostic is emitted.

Comment (DenisBakhvalov) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: I'm OK with both error and warning options, but as @elizabethandrews pointed out, let's keep it as a warning for consistency.

@AaronBallman, how I can reach the AST that you showed for the example below?
```
[[intel::sycl_esimd_vectorize(8)]] __attribute__((sycl_device)) void foo3();
[[intel::sycl_esimd_vectorize(16)]] __attribute__((sycl_device)) void foo3() {}
```

AFAIK, the attribute for the definition of `foo3` is not `Inherited`. It would be `Inherited` if the definition of `foo3` would not have the attribute specified. So, how I could make the definition of `foo3` inherit the attribute from the declaration? Should I examine the previous declaration of `foo3` and skip adding (in `AddSYCLIntelESimdVectorizeAttr`) the attribute if it exists in the declaration already? Would appreciate examples of other attributes which I can copy from.

Comment (AaronBallman) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: > > What I'd expect is that a conflict between values would be an error rather than a warning, to be honest (that's how most attributes work).
> 
> For SYCL attributes, I think conflicting values currently emit a warning. If we want to change this, this should be changed for the existing attributes also right?

Agreed, and I'm not suggesting a change from warning to error in this patch. I went back and looked again and community is more inconsistent than I had realized, so "that's how most attributes work" was incorrect.

Comment (AaronBallman) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: > I'm OK with both error and warning options, but as @elizabethandrews pointed out, let's keep it as a warning for consistency.

Agreed.

> @AaronBallman, how I can reach the AST that you showed for the example below?
> 
> ```
> [[intel::sycl_esimd_vectorize(8)]] __attribute__((sycl_device)) void foo3();
> [[intel::sycl_esimd_vectorize(16)]] __attribute__((sycl_device)) void foo3() {}
> ```
> 
> AFAIK, the attribute for the definition of `foo3` is not `Inherited`. It would be `Inherited` if the definition of `foo3` would not have the attribute specified. So, how I could make the definition of `foo3` inherit the attribute from the declaration? Should I examine the previous declaration of `foo3` and skip adding (in `AddSYCLIntelESimdVectorizeAttr`) the attribute if it exists in the declaration already? Would appreciate examples of other attributes which I can copy from.

Oh, interesting, you're right, all of our attributes are doing the right/wrong thing here.

We have a `-Wignored-attributes` warning when there's a duplicate attribute, but we continue to use its value (the diagnostic gives the wrong impression), but the natural consequence of the way we structure our code is to leave the attribute in place on the declaration in which it was found. (`AddSYCLIntelESimdVectorizeAttr()` is called twice, once for each declaration, `MergeSYCLIntelESimdVectorizeAttr()` is called once for the redeclaration but doesn't remove the already-added attribute.)

So I don't think there's anything for you to change here. If we did, it'd be inconsistent with our other attributes, and if we changed those, we'd risk silently changing program behavior for users who are relying on the duplicate attribute behavior. If we're going to break people, I'd rather it be done by turning this into an error rather than a warning. More thought it needed as to what the right thing to do here is, but thankfully, this is pretty low priority as it's an edge case.

Comment (AaronBallman) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: tl;dr: I'm back to LGTM. :-)

Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: @AaronBallman based on what you said, the latest commit in this PR should be reverted right?

Comment (AaronBallman) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: Yes -- I had missed that there was another commit, but I spotted it now. Good catch, @elizabethandrews!

Comment (DenisBakhvalov) [reply clang/test/CodeGenSYCL/esimd-vectorize-md.cpp:12]: Sure. Thanks, @AaronBallman and @elizabethandrews, I will go ahead and revert the last commit.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Review by kbobrovs — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]
